src/azure/azureController.ts

```diff
@@ -83,4 +91,91 @@ export class AzureController {
         this.azureMessageDisplayer = new AzureMessageDisplayer();
     }
 
+    public async getTokens(profile: ConnectionProfile, accountStore: AccountStore, settings: AADResource): Promise<ConnectionProfile> {
+        let account: IAccount;
+        let config = vscode.workspace.getConfiguration('mssql').get('azureActiveDirectory');
+        if (config === utils.azureAuthTypeToString(AzureAuthType.AuthCodeGrant)) {
+            let azureCodeGrant = await this.createAuthCodeGrant();
+            account = await azureCodeGrant.startLogin();
+            await accountStore.addAccount(account);
+            const token = await azureCodeGrant.getAccountSecurityToken(
+                account, azureCodeGrant.getHomeTenant(account).id, settings
+            );
+            profile.azureAccountToken = token.token;
+            profile.email = account.displayInfo.email;
+            profile.accountId = account.key.id;
+        } else if (config === utils.azureAuthTypeToString(AzureAuthType.DeviceCode)) {
+            let azureDeviceCode = await this.createDeviceCode();
+            account = await azureDeviceCode.startLogin();
+            await accountStore.addAccount(account);
+            const token = await azureDeviceCode.getAccountSecurityToken(
+                account, azureDeviceCode.getHomeTenant(account).id, settings
+            );
+            profile.azureAccountToken = token.token;
+            profile.email = account.displayInfo.email;
+            profile.accountId = account.key.id;
+        }
+        return profile;
+    }
+
+    public async refreshTokenWrapper(profile, accountStore, accountAnswer, settings: AADResource): Promise<ConnectionProfile> {
+        let account = accountStore.getAccount(accountAnswer.key.id);
+        if (!account) {
+            throw new Error('account not found');
+        }
+        if (account.isStale) {
+            accountStore.removeAccount(account.key.id);
+            throw new Error('account is stale. Please re-do login');
```

Comment:
You might want to throw these errors as information messages to the user so they know what's going on. Also, localize in that case. All strings that are visible in the UI need to be localized. Internal strings are fine.